flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/HashSubpartitionBufferAccumulator.java

```diff
@@ -68,17 +83,32 @@ public HashSubpartitionBufferAccumulator(
     // ------------------------------------------------------------------------
 
     public void append(ByteBuffer record, Buffer.DataType dataType) throws IOException {
-        if (dataType.isEvent()) {
-            writeEvent(record, dataType);
-        } else {
-            writeRecord(record, dataType);
+        synchronized (lock) {
+            if (isReleased) {
+                return;
+            }
+            if (dataType.isEvent()) {
+                writeEvent(record, dataType);
+            } else {
+                writeRecord(record, dataType);
+            }
         }
     }
 
+    /** This method can only be called by the task thread. */
     public void close() {
+        recycleBuffers();
         checkState(unfinishedBuffers.isEmpty(), "There are unfinished buffers.");
     }
 
+    public void release() {
```

Comment:
Maybe we only need release the buffers in close method and get rid of this lock in the same time.

Reply:
@reswqa Good catch! This idea can significantly simplify the fix.